### PR TITLE
Do not use global variables in Vmbytegen.

### PR DESCRIPTION
**Pull request by ppedrot**
In addition to be bad practice and thread-unsafe, which are just theoretical considerations, this is also a source of a memory leak in some developments. Indeed, the potentially huge bytecode stays reachable way beyond the point where it is not necessary anymore.

### CODE REVIEW

**Comment by silene** [comment]
I have a hard time convincing myself that all the stack size changes are actually correct. Do we have a simple way to check that the `.vo` files (and thus the generated bytecode) are left unchanged by this pull request?

File: kernel/vmbytegen.ml [hunk 1]
```diff
@@ -629,38 +631,35 @@ let rec compile_lam env cenv lam sz cont =
...
+        let () = set_max_stack_size env_body (arity + 4) in
+        let cont = compile_lam env env_body body (arity+1) (cont_cofix arity) in
+        let cont = ensure_stack_capacity env_body cont in
         lbl_bodies.(i) <- lbl;
-        fun_code := Ksequence (add_grab (arity+1) lbl cont) :: !fun_code;
+        push_fun env (add_grab (arity+1) lbl cont)
       done;
       let fv = !rfv in
-      set_max_stack_size (sz + fv.size + ndef + 2);
+      let () = set_max_stack_size cenv (sz + fv.size + ndef + 2) in
```
Comment (maximedenes) [inline kernel/vmbytegen.ml:656]: Isn't this still a sequence? Why change it to a let binding?

Comment (ppedrot) [reply kernel/vmbytegen.ml:656]: I personally hate the sequence operator in OCaml, due to parsing weirdness it's way to easy to break with innocuous code manipulations. Furthermore, this way of writing things makes it very explicit that we're doing this for the side-effect.

**Comment by ppedrot** [comment]
@silene since the vo files embed the hash of the OCaml plugins, it's hard to check for byte-for-byte equality, as it won't hold in general. I tried to preserve the previous behaviour for max stack depth computation, but even currently on master there is some weirdness going on, e.g. with non-constant branches of cases.

**Comment by ppedrot** [comment]
Weirdly enough this doesn't solve the specific memory consumption I was looking at, but makes it even worse as the memory allocated by the bytecode is never reclaimed. Even more weirdly, just setting the fun_code reference at the exit of the VM compilation still prevents the bytecode memory from being reclaimed. Either memtrace is lying to me or there is something very wrong going on there.

**Comment by maximedenes** [comment]
How about opening an issue with your findings and a repro case, so that we don't forget about it?

**Comment by ppedrot** [comment]
That's a bit hard because it requires a custom branch, but when #16827 lands it should be easier to do.

**Comment by silene** [comment]
> Even more weirdly, just unsetting the fun_code reference at the exit of the VM compilation still prevents the bytecode memory from being reclaimed. Either memtrace is lying to me or there is something very wrong going on there.

Keep in mind that the main culprit when it comes to the VM leaking memory is `coq_tcode_of_code` with its use of `coq_stat_alloc(len)`. Not allocating on the OCaml heap is critical for correctness, but it also means that we should understand when we can really free it. (And I am not sure it is as simple as putting a finalizer on the corresponding abstract block.)

**Comment by ppedrot** [comment]
I agree that leaking the tcode is bad, but this is not what I'm witnessing here. The leak under consideration is generated by the OCaml data related to Vmbytecodes.instruction, and is not a big chunk of C memory.

**Comment by silene** [comment]
The other thing that comes to mind is the bytecode stack. You could try to print `coq_sp` at the entry of `coq_interprete` and see whether it keeps decreasing, which would mean that some things are kept alive indefinitely.

**Comment by silene** [comment]
I support the motivation for this patch, but I honestly have a very hard time reviewing it. Indeed, the patch contains a lot of small changes that deeply modify the semantics of the code.

For example, `let env_type = comp_env_cofix_type ndef rfv in` was moved from outside the loop to the loop body. So, this mutated structure is no longer shared through all the loop iterations. Perhaps we do not care. Perhaps it fixes a bug in the old code. But perhaps it actually introduces a new bug. I just do not know.

Since the patch does not bring the expected performance improvements, it might be safer to just ignore it.

**Comment by ppedrot** [comment]
@silene I've dug a bit into the compiled code and compared hashes of the generated bytecode across versions. Indeed the PR was generating in some corner cases slightly different max stack size for case nodes, but it's unclear to me that the behaviour on master is really specified (or understood at all).

I have restored the old behaviour by passing an int reference instead of a mutable field, which matters for the few places where we do not wrap the call to code generation inside ensure_stack_capacity. This is notably the case for the call to push_param in the Case node compilation, where we have
```ocaml
        let code_b =
          compile_lam env (push_param arity sz_b cenv)
            body (sz_b+arity) (add_pop arity (branch::!c)) in
```
This was the one generating different max_stack_size as the result survives the for loop and is passed by side-effect in
```ocaml
      let annot =
        {rtbl = rtbl; tailcall = is_tailcall;
         Vmvalues.max_stack_size = cenv.max_stack_size.contents - sz}
      in
```
I *think* this is wrong because the branches we consider are always non-constant and thus encoded as closures, so the stack capacity of these should be handled when entering these closures. At worst it's an overapproximation so it's still safe, but this means we could be making the stack grow for no good reason when evaluating these pattern-matchings.

With the current version of the PR, I have bit-for-bit compatibility of all bytecode generated from the standard library. While not a formal proof, I think it is reasonable to believe that this PR is now equivalent to the previous code and therefore ready to merge.

**Comment by ppedrot** [comment]
Ping @silene to check whether you're fine with the current code.

**Comment by silene** [comment]
> I think this is wrong because the branches we consider are always non-constant and thus encoded as closures, so the stack capacity of these should be handled when entering these closures.

That is not how it works. Branches are compiled as if they were closures, but they are not executed as closures. In particular, branches are executed in the activation frame of the current function. For example, if the code of a branch references a variable outside the match, then it directly accesses its value on the stack of the enclosing function; the value was not first copied into the closure environment of the branch (which would be the case if it was an actual closure). This is critical for performance. Another example is when the match is in a terminal position; you do not even want to waste time doing a tail call; you just want to immediately execute the branch.

Anyway, this means that the stack needs to be large enough to account for whichever branch might later be taken. So, this is indeed a bit pessimistic, as it could cause a stack reallocation even though the whole reduction might be about to terminate without ever needing an extra stack reallocation. But it is better than checking the stack capacity whenever you enter a match.

> With the current version of the PR, I have bit-for-bit compatibility of all bytecode generated from the standard library.

Good enough.

**Review by silene — APPROVED** [review]

**Comment by ppedrot** [comment]
@silene you should probably assign then.

**Pull request merged**